homeassistant/components/smart_meter_texas/get_ssl_context.py

```diff
@@ -0,0 +1,57 @@
+import logging
+import re
+import socket
+import ssl
+import urllib
+
+import OpenSSL.crypto as crypto
+import certifi
+from smart_meter_texas.const import BASE_HOSTNAME
+
+_LOGGER = logging.getLogger(__name__)
+
+def get_ssl_context():
+    """Creates a usable SSL Context for Smart Meter Texas.
+    This will attempt to download and install the CA Issuers certificate into the SSL Context, then re-enable the SSL checking before proceeding.
+    """
+    ssl_context = None
+    try:
+        caiKey = "CA Issuers - URI:"
+        re_issuers_uri = re.compile(r"(https?://+[\w\d:#@%/;$()~_?\+-=\\\.&]*)", re.UNICODE)
```

Comment:
This seem brittle as it relies on extracting undecoded ASN.1 data.  We should decode the data and get the url instead

Reply:
I tried a few different approaches but wasn't able to figure out how to just get the CA Issuers URI from the previously decoded asn1.

Reply:
Using the asn.1 example, something like this might work:

```python3
import ssl, socket
import OpenSSL.crypto as crypto
import asn1
import sys


tag_id_to_string_map = {
    asn1.Numbers.Boolean: "BOOLEAN",
    asn1.Numbers.Integer: "INTEGER",
    asn1.Numbers.BitString: "BIT STRING",
    asn1.Numbers.OctetString: "OCTET STRING",
    asn1.Numbers.Null: "NULL",
    asn1.Numbers.ObjectIdentifier: "OBJECT",
    asn1.Numbers.PrintableString: "PRINTABLESTRING",
    asn1.Numbers.IA5String: "IA5STRING",
    asn1.Numbers.UTCTime: "UTCTIME",
    asn1.Numbers.Enumerated: "ENUMERATED",
    asn1.Numbers.Sequence: "SEQUENCE",
    asn1.Numbers.Set: "SET",
}


class_id_to_string_map = {
    asn1.Classes.Universal: "U",
    asn1.Classes.Application: "A",
    asn1.Classes.Context: "C",
    asn1.Classes.Private: "P",
}


def tag_id_to_string(identifier):
    """Return a string representation of a ASN.1 id."""
    if identifier in tag_id_to_string_map:
        return tag_id_to_string_map[identifier]
    return "{:#02x}".format(identifier)


def class_id_to_string(identifier):
    """Return a string representation of an ASN.1 class."""
    if identifier in class_id_to_string_map:
        return class_id_to_string_map[identifier]
    raise ValueError("Illegal class: {:#02x}".format(identifier))


def value_to_string(tag_number, value):
    if tag_number == asn1.Numbers.ObjectIdentifier:
        return value
    elif isinstance(value, bytes):
        return "0x" + str(binascii.hexlify(value).upper())
    elif isinstance(value, str):
        return value
    else:
        return repr(value)


def pretty_print(input_stream, output_stream, indent=0):
    """Pretty print ASN.1 data."""
    while not input_stream.eof():
        tag = input_stream.peek()
        if tag.typ == asn1.Types.Primitive:
            tag, value = input_stream.read()
            output_stream.write(" " * indent)
            output_stream.write(
                "[{}] {}: {}\n".format(
                    class_id_to_string(tag.cls),
                    tag_id_to_string(tag.nr),
                    value_to_string(tag.nr, value),
                )
            )
        elif tag.typ == asn1.Types.Constructed:
            output_stream.write(" " * indent)
            output_stream.write(
                "[{}] {}\n".format(
                    class_id_to_string(tag.cls), tag_id_to_string(tag.nr)
                )
            )
            input_stream.enter()
            pretty_print(input_stream, output_stream, indent + 2)
            input_stream.leave()


ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE
with ctx.wrap_socket(socket.socket(), server_hostname="smartmetertexas.com") as s:
    s.connect(("smartmetertexas.com", 443))
    cert_bin = s.getpeercert(True)
    x509 = crypto.load_certificate(crypto.FILETYPE_ASN1, cert_bin)
    for idx in range(x509.get_extension_count()):
        ext = x509.get_extension(idx)
        short_name = ext.get_short_name()
        if short_name == b"authorityInfoAccess":
            decoder = asn1.Decoder()
            print(str(ext))
            data = ext.get_data()
            decoder.start(data)
            pretty_print(decoder, sys.stdout)
```

Reply:
Thanks! I was able to get something working by narrowing it down to this:

```
def value_to_string(tag_number, value):
    if tag_number == asn1.Numbers.ObjectIdentifier:
        return value
    elif isinstance(value, bytes):
        return "0x" + str(binascii.hexlify(value).upper())
    elif isinstance(value, str):
        return value
    else:
        return repr(value)

def find_ca_issuers_uri(input_stream, tag_ca_issuers_uri_found = False):
    """Lookup the CA Issuers - URI Object and return the value."""
    ca_issuers_uri = None
    while not input_stream.eof() and not ca_issuers_uri:
        tag = input_stream.peek()
        if tag.typ == asn1.Types.Primitive:
            tag, value = input_stream.read()
            # If we've located the tag, extract the URI value
            if (tag_ca_issuers_uri_found):
                str_value = value_to_string(tag.nr, value)
                if (str_value):
                    ca_issuers_uri = str_value.decode('utf-8')
                    tag_ca_issuers_uri_found = False
                    break
                else:
                    tag_ca_issuers_uri_found = False
            # Locate the tag for CA Issuers - URI
            elif (value_to_string(tag.nr, value) == '1.3.6.1.5.5.7.48.2'):
                tag_ca_issuers_uri_found = True

        elif tag.typ == asn1.Types.Constructed:
            input_stream.enter()
            ca_issuers_uri = find_ca_issuers_uri(input_stream)
            input_stream.leave()

    return ca_issuers_uri

async def main():
    ca_issuers_uri = None
    context = ssl.create_default_context(capath=certifi.where())
    context.check_hostname = False
    context.verify_mode = ssl.CERT_NONE
    with context.wrap_socket(socket.socket(), server_hostname=BASE_HOSTNAME) as s:
        s.connect((BASE_HOSTNAME, 443))
        cert_bin = s.getpeercert(True)
        x509 = crypto.load_certificate(crypto.FILETYPE_ASN1, cert_bin)
        for idx in range(x509.get_extension_count()):
            ext = x509.get_extension(idx)
            short_name = ext.get_short_name()
            if short_name == b"authorityInfoAccess":
                decoder = asn1.Decoder()
                data = ext.get_data()
                decoder.start(data)
                ca_issuers_uri = find_ca_issuers_uri(decoder)

    print(ca_issuers_uri)

```